frontend/components/TodoList.js

```diff
@@ -0,0 +1,117 @@
+import React, { useEffect, useState } from "react";
+import { Heading, HStack, Switch, VStack, Text, ScrollView } from "native-base";
+import Task from "./Task";
+import TodoCategoryGroup from "./TodoCategoryGroup";
+
+import { TODO_API } from "../api/todo.api";
+import { TASK_API } from "../api/task.api";
+
+const TodoList = (props) => {
+  const [data, setData] = useState(DUMMY_TODOS);
+  const [isGrouped, setIsGrouped] = useState(false);
+
+  const toggleGrouping = () => {
+    setIsGrouped((current) => !current);
+  };
+
+  const fetchTodoList = async () => {
+    const data = await TODO_API.fetchTodoList();
+    setData((current) => {
+      const isGrouped = !current.isGrouped;
+      return { isGrouped, data };
+    });
+  };
+
+  useEffect(() => {
+    // fetchTodoList();
+  }, []);
+
+  const groupedTasks = () => {
+    const obj = data.reduce((groups, task) => {
+      const category = task.category || "None";
+      groups[category] = [...(groups[category] || []), task];
+      return groups;
+    }, {});
+    return Object.entries(obj);
+  };
+
+  const removeTask = (taskId) => {
+    setData((data) => data.filter((task) => task.id !== taskId));
+    TODO_API.removeFromTodo(taskId);
+  };
+
+  const completeTask = (taskId) => {
+    setData((data) => data.filter((task) => task.id !== taskId));
```

Comment:
Line 44 causes a bug. Completed task is removed and the task of its index (the one after it) is marked completed. Bug has no effect if tasks are checked bottom-up. When removing line 44, bug is fixed

Reply:
I ran into a problem like this, make sure to when tasks have unique IDs, and also if you're incrementing IDs manually increment it and store it in a variable and then change it's state using setID() or whatever.

Reply:
Actually the problem wasn't with the removing logic, the data was fine.

The problem was that I used the `index` of the task in the array as a key, so when removing an item with index 0 the next item index becomes 0, and React considers both the same since they have the same key and propagates the state to the next item.
https://beta.reactjs.org/learn/rendering-lists#why-does-react-need-keys